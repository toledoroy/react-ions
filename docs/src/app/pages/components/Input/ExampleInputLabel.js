import React from 'react'
import Input from 'react-ions/lib/components/Input'

const ExampleInputLabel = () => (
  <Input value='' label='This is the label' helperText='Helper text.' />
)

export default ExampleInputLabel
