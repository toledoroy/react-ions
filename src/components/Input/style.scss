@import '../../styles/common/colors';
@import '../../styles/common/screen';

.input-component {
  position: relative;

  label {
    display: block;
    margin-bottom: 10px;
    font-size: 16px;
    color: $navy-text;
    font-weight: 200;
  }

  .input-container {
    align-items: center;
    display: flex;
    position: relative;

    .flex-wrapper {
      align-items: center;
      display: flex;
      height: 100%;
      line-height: 0;
      position: absolute;
      top: 0; // for IE 11 :)

      &.suffix {
        // compensates for the 16px margin added below
        right: -16px;
      }
    }

    .prefix, .suffix {
      color: $primary-4;
    }

    .prefix {
      margin-left: 16px;
    }

    .suffix {
      margin-right: 16px;
    }

    input {
      color: $primary-4;
      border: solid 2px $neutral-3;
      border-radius: 3px;
      box-sizing: border-box;
      font-size: 14px;
      padding: 10px;
      width: 100%;
      transition: border-color ease-in-out .15s;

      &:disabled {
        background-color: rgba($neutral-3, .6);
        cursor: not-allowed;

        &:hover {
          border: solid 2px $neutral-3;
        }
      }

      &.fill {
        width: 100%;
      }
    }
  }

  &.input-error {
    input {
      border: 2px solid $danger;

      &:focus {
        outline: 0;
      }  
    }
  }

  &:not(.input-error) {
    input:hover {
      border-color: rgba($neutral-4, .4);
    }

    input:focus {
      border-color: rgba($neutral-4, .6);
      outline: 0;
    }
  }

}
