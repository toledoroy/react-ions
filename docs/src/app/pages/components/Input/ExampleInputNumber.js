import React from 'react'
import Input from 'react-ions/lib/components/Input'

class ExampleInputNumber extends React.Component {
  constructor(props) {
    super(props)
  }

  state = {
    statusNumber: '',
    statusString: '',
    valueNumber: 12.34,
    valueString: '12.34'
  }

  UNSAFE_componentWillMount = () => {
    this.setState({statusNumber: 'My value is: ' + this.state.valueNumber + ' and it is of type: ' + typeof this.state.valueNumber})
    this.setState({statusString: 'My value is: ' + this.state.valueString + ' and it is of type: ' + typeof this.state.valueString})
  }

  handleChangeNumber = event => {
    if (event.target.value === '') {
      event.target.value = '\'\' (empty string)'
    }
    this.setState({statusNumber: 'My value is: ' + event.target.value + ' and it is of type: ' + typeof event.target.value})
  }

  handleChangeString = event => {
    if (event.target.value === '') {
      event.target.value = '\'\' (empty string)'
    }
    this.setState({statusString: 'My value is: ' + event.target.value + ' and it is of type: ' + typeof event.target.value})
  }

  render = () => {
    return (
      <div>
        <Input valueType='number' value={this.state.valueNumber} changeCallback={this.handleChangeNumber} label='Input of number type' />
        <code>{this.state.statusNumber}</code>
        <Input valueType='string' value={this.state.valueString} changeCallback={this.handleChangeString} label='Input of string type' />
        <code>{this.state.statusString}</code>
      </div>
    )
  }
}

export default ExampleInputNumber
