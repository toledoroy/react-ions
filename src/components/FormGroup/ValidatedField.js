import React from 'react'
import { array, object, string } from 'prop-types'
import optclass from '../internal/OptClass'
import style from './style.scss'

const ValidatedField = WrappedComponent => {
  return class extends React.Component {
    static propTypes = {
      error: string
    }

    render = () => {
      const { className, ...fieldProps } = this.props
      const wrapperErrorClass = fieldProps.error ? 'has-error' : ''
      const validatedFieldClass = optclass(style, ['validated-field', wrapperErrorClass, className])

      return (
        <div className={validatedFieldClass}>
          <WrappedComponent {...fieldProps} />
          {fieldProps.error &&
            <span className={style['has-error__message']}>
              {fieldProps.error}
            </span>
          }
        </div>
      )
    }
  }
}

export default ValidatedField
