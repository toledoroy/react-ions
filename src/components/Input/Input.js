import React from 'react'
import PropTypes from 'prop-types'
import optclass from '../internal/OptClass'
import style from './style.scss'

/**
 * The Input component.
 */
class Input extends React.Component {
  constructor(props) {
    super(props)
  }

  static defaultProps = {
    disabled: false,
    value: '',
    valueType: 'string'
  }

  static propTypes = {
    /**
     * A class name to be used for local styles or integrations (required to support styled-components)
     */
    className: PropTypes.string,
    /**
     * Whether the input is disabled.
     */
    disabled: PropTypes.bool,
    /**
     * Disallow the user from editing the input.
     */
    readOnly: PropTypes.bool,
    /**
     * Text shown above the input.
     */
    label: PropTypes.string,
    /**
     * Value of the input.
     */
    value: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.string
    ]),
    /**
     * Type of the value.
     */
    valueType: PropTypes.oneOf(['string', 'number']),
    /**
     * Optional placeholder text.
     */
    placeholder: PropTypes.string,
    /**
     * Name of the input.
     */
    name: PropTypes.string,
    /**
     * Name of the input.
     */
    prefix: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.node
    ]),
    /**
     * Optional prefix to add to the input.
     */
    suffix: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.node
    ]),
    /**
     * Optional suffix to add to the input.
     */
    optClass: PropTypes.string,
    /**
     * A callback function to be called when the input changes.
     */
    changeCallback: PropTypes.func,
    /**
     * A callback function to be called when the input is focused.
     */
    focusCallback: PropTypes.func,
    /**
     * A callback function to be called when the input is blurred.
     */
    blurCallback: PropTypes.func,
    /**
     * A callback function to be called when the input is clicked.
     */
    onClick: PropTypes.func,
    /**
     * A callback function to be called when the onkeyup event is fired.
     */
    onKeyUp: PropTypes.func,
    /**
     * A callback function to be called when the onkeypress event is fired.
     */
    onKeyPress: PropTypes.func,
    /**
     * A callback function to be called when the onkeydown event is fired.
     */
    onKeyDown: PropTypes.func,
    /**
     * A helper will render inline style='width: <value>'.
     */
    width: PropTypes.string,
    /**
     * A fallback value for when the value is null.
     */
    nullValue: PropTypes.string,
    /**
     * A string to be used as the ID.
     */
    id: PropTypes.string
  }

  _getValue = props => {
    return (props.value === null && typeof props.nullValue !== 'undefined') ? props.nullValue : props.value
  }

  componentWillMount = () => {
    this.setState({ value: this._getValue(this.props) })
  }

  componentDidMount = () => {
    this.handleInlineStyles()
  }

  componentWillReceiveProps = nextProps => {
    if (nextProps.value !== this.props.value) {
      this.setState({ value: this._getValue(nextProps) })
    }
  }

  componentDidUpdate = prevProps => {
    if (prevProps.suffix !== this.props.suffix || prevProps.prefix !== this.props.prefix) {
      this.handleInlineStyles()
    }
  }

  handleInlineStyles = () => {
    const inputStyles = {}

    if (this.props.prefix) {
      // Add 24 to accommodate for left and right padding of prefix (16+8)
      inputStyles.paddingLeft = this._prefix.getBoundingClientRect().width + 24
    }
    if (this.props.suffix) {
      // Add 24 to accommodate for left and right padding of prefix (8+16)
      inputStyles.paddingRight = this._suffix.getBoundingClientRect().width + 24
    }

    this.setState({ inputStyles })
  }

  handleChange = event => {
    event.persist()
    const value = (this.props.valueType === 'number' && event.target.value !== '' && !isNaN(event.target.value))
                  ? parseFloat(event.target.value) : event.target.value

    this.setState({value: event.target.value}, () => {
      this.props.changeCallback && this.props.changeCallback({ target: { name: this.props.name, value } })
    })
  }

  handleFocus = event => {
    this.props.focusCallback && this.props.focusCallback(event)
  }

  handleBlur = event => {
    this.props.blurCallback && this.props.blurCallback(event)
  }

  focus = () => {
    this._input.focus()
  }

  render = () => {
    const {prefix, suffix, label, optClass, className, width, disabled} = this.props

    const disabledClass = disabled ? style['input-disabled'] : null
    const widthStyle = width ? { width: width } : null
    const prefixClass = prefix ? style['prefix'] : null
    const suffixClass = suffix ? style['suffix'] : null
    const inputClass = optclass(style, ['input-component', className, optClass, disabledClass, prefixClass, suffixClass])
    const inputContainerClass = style['input-container']
    const flexWrapperClass = optclass(style, 'flex-wrapper')
    const flexWrapperClassSuffix = optclass(style, [flexWrapperClass, 'suffix'])

    return (
      <div className={inputClass}>
        {label && <label>{label}</label>}
        <div className={inputContainerClass} style={widthStyle}>
          {prefix && <div className={flexWrapperClass}>
            <div ref={c => this._prefix = c} className={prefixClass}>{prefix}</div>
          </div>}

          <input
            id={this.props.id}
            ref={c => this._input = c}
            value={this.state.value}
            onFocus={this.handleFocus}
            onChange={this.handleChange}
            onClick={this.props.onClick}
            onBlur={this.handleBlur}
            disabled={this.props.disabled}
            readOnly={this.props.readOnly}
            placeholder={this.props.placeholder}
            style={this.state.inputStyles}
            onKeyUp={this.props.onKeyUp}
            onKeyPress={this.props.onKeyPress}
            onKeyDown={this.props.onKeyDown}>
          </input>

          {suffix && <div className={flexWrapperClassSuffix}>
            <div ref={c => this._suffix = c} className={suffixClass}>{suffix}</div>
          </div>}
        </div>
      </div>
    )
  }
}

export default Input
