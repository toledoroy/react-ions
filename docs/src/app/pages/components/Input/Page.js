import React from 'react'
import PropsList from 'private/modules/PropsList'
import docs from '!!docgen!react-ions/lib/components/Input/Input'
import CodeExample from 'private/modules/CodeExample'
import styles from 'private/css/content'
import ExampleInputDefault from './ExampleInputDefault'
import exampleInputDefaultCode from '!raw!./ExampleInputDefault'
import ExampleInputPlaceholder from './ExampleInputPlaceholder'
import exampleInputPlaceholderCode from '!raw!./ExampleInputPlaceholder'
import ExampleInputDisabled from './ExampleInputDisabled'
import exampleInputDisabledCode from '!raw!./ExampleInputDisabled'
import ExampleInputError from './ExampleInputError'
import exampleInputErrorCode from '!raw!./ExampleInputError'
import ExampleInputLabel from './ExampleInputLabel'
import exampleInputLabel from '!raw!./ExampleInputLabel'
import ExampleInputCallback from './ExampleInputCallback'
import exampleInputCallbackCode from '!raw!./ExampleInputCallback'
import ExampleInputPopulated from './ExampleInputPopulated'
import exampleInputPopulatedCode from '!raw!./ExampleInputPopulated'
import ExampleInputNumber from './ExampleInputNumber'
import exampleInputNumberCode from '!raw!./ExampleInputNumber'
import ExampleInputPrefixSuffix from './ExampleInputPrefixSuffix'
import exampleInputPrefixSuffixCode from '!raw!./ExampleInputPrefixSuffix'
import ExampleInputCopy from './ExampleInputCopy'
import exampleInputCopyCode from '!raw!./ExampleInputCopy'

const description = {
  inputDefault: 'This is the `input component` as it appears by default.',
  inputPlaceholder: 'This is the `input component` with placeholder text.',
  inputDisabled: 'This is the disabled `input component`.',
  inputError: 'This is the `input component` with error.',
  inputWithLabel: 'This is the `input component` with a label and helper text.',
  inputCallback: 'This is the `input component` with a callback function. __Note__: the `style import` and `code` tag is for display purposes only.',
  inputNumber: 'This is the `input component` with different types.',
  inputPrefixSuffix: 'This is the `input component` with a variety of prefix and suffix.',
  inputCopy: 'This is the `input component` with click to copy.'
}

const InputPage = () => (
  <div>
    <div className={styles.content}>
      <div className={styles.block}>
        <CodeExample
          title='Default Input'
          description={description.inputDefault}
          markup={exampleInputDefaultCode}>
          <ExampleInputDefault />
        </CodeExample>
        <CodeExample
          title='Input with Placeholder Text'
          description={description.inputPlaceholder}
          markup={exampleInputPlaceholderCode}>
          <ExampleInputPlaceholder />
        </CodeExample>
        <CodeExample
          title='Disabled Input'
          description={description.inputDisabled}
          markup={exampleInputDisabledCode}>
          <ExampleInputDisabled />
        </CodeExample>
        <CodeExample
          title='Error Input'
          description={description.inputError}
          markup={exampleInputErrorCode}>
          <ExampleInputError />
        </CodeExample>
        <CodeExample
          title='Input with Label and Helper Text'
          description={description.inputWithLabel}
          markup={exampleInputLabel}>
          <ExampleInputLabel />
        </CodeExample>
        <CodeExample
          title='Input with callback function(s)'
          description={description.inputCallback}
          markup={exampleInputCallbackCode}>
          <ExampleInputCallback />
        </CodeExample>
        <CodeExample
          title='Handling updates via props'
          description={description.inputCallback}
          markup={exampleInputPopulatedCode}>
          <ExampleInputPopulated />
        </CodeExample>
        <CodeExample
          title='Input with different types'
          description={description.inputNumber}
          markup={exampleInputNumberCode}>
          <ExampleInputNumber />
        </CodeExample>
        <CodeExample
          title='Input with a variety of prefix and suffix'
          description={description.inputPrefixSuffix}
          markup={exampleInputPrefixSuffixCode}>
          <ExampleInputPrefixSuffix />
        </CodeExample>
        <CodeExample
          title='Input with click to copy'
          description={description.inputCopy}
          markup={exampleInputCopyCode}>
          <ExampleInputCopy id='test-input-copy' value='Copy me to your clipboard' />
        </CodeExample>
        <div className={styles.block}>
          <h3>Props</h3>
          <PropsList list={docs[0].props} />
        </div>
      </div>
    </div>
  </div>
)

export default InputPage
