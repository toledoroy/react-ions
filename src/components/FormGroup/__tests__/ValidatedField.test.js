import React from 'react'
import ValidatedField from '../ValidatedField'
import Input from '../../Input'

describe('ValidatedField', () => {
  let wrapper

  const ValidatedInput = ValidatedField(Input)

  it('should shallow render itself', () => {
    expect(shallow(<ValidatedInput error='Error here.' />)).toMatchSnapshot()
  })

  it('should have a custom class', () => {
    expect(shallow(<ValidatedInput className='custom-class' />)).toMatchSnapshot()
  })
})
