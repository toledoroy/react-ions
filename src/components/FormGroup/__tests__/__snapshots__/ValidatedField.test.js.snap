// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidatedField should have a custom class 1`] = `
<div
  className="validated-field custom-class"
>
  <Input
    disabled={false}
    type="text"
    value=""
    valueType="string"
  />
</div>
`;

exports[`ValidatedField should shallow render itself 1`] = `
<div
  className="validated-field has-error"
>
  <Input
    disabled={false}
    error="Error here."
    type="text"
    value=""
    valueType="string"
  />
  <span
    className="has-error__message"
  >
    Error here.
  </span>
</div>
`;
